src/main/java/com/vlf/intentsender/ui/MainToolWindowPresenter.kt

```diff
@@ -0,0 +1,295 @@
+    }
+
+    override fun onFlagsClicked() {
+        view.showFlagsSelection(listOf(*IntentFlags.values()), selectedFlags)
+    }
+
+    override fun onFlagsSelected(newFlags: List<IntentFlags>) {
+        if (newFlags.isEmpty()) {
+            selectedFlags = listOf(IntentFlags.NONE)
+        }
+
+        val flags = mutableListOf<IntentFlags>()
+        flags.addAll(newFlags)
+        // Remove NONE if it was added
+        if (flags.size > 1
+            && flags.contains(IntentFlags.NONE)
+        ) {
+            flags.remove(IntentFlags.NONE)
+        }
+        selectedFlags = flags
+
+        view.displaySelectedFlags(selectedFlags)
+    }
+
+    private inner class SendAdbCommandWorker(
+        private val command: Command
+    ) : SwingWorker<String?, String?>() {
+        override fun doInBackground(): String? {
+            var error: String? = null
+            lastCommandOutput = null
+            try {
+                AdbHelper.getInstance().setOutputListener { output ->
+                    lastCommandOutput = output
+                }
+                error = AdbHelper.getInstance().sendCommand(command, selectedDevice)
+            } catch (e: Exception) {
+                e.printStackTrace()
+            }
+            return error
+        }
+
+        override fun done() {
+            val error: String? = try {
+                get()
+            } catch (e: Exception) {
+                if (e.message != null) e.message else ERROR_UNKNOWN
+            }
+            if (error == null) {
+                handleCommandExecutionSuccess(command)
+            } else {
+                handleCommandExecutionError(error, command)
+            }
+            view.enableStartButtons(true)
+        }
+
+        private fun handleCommandExecutionError(error: String, command: Command) {
+            println("Sending command FAILED: $error")
+            view.showCommandExecutionError(error)
+        }
+
+        private fun handleCommandExecutionSuccess(command: Command) {
+            println("SUCCESS sending command")
+            HistoryUtils.saveCommand(command)
+            view.showCommandSentSuccessfully()
+        }
+    }
+
+    private inner class RestartAdbWorker(
+        private val adbHelper: AdbHelper
+    ) : SwingWorker<Void?, Void?>() {
+
+        override fun doInBackground(): Void? {
+            adbHelper.restartAdb()
+            return null
+        }
+
+        override fun done() {
+            // TODO(vfarafonov, 12/23/20): move Done actions to constructor as lambda.
+            view.setAdbInitProgressIndicatorVisible(false)
+            view.setLocateAdbButtonVisible(false)
+            view.setIntentCreationLayoutVisible(true)
+
+            onUpdateDevicesClicked()
+        }
+    }
+
+    private inner class DevicesListener : IDeviceChangeListener {
+
+        override fun deviceConnected(iDevice: IDevice) {
+            SwingUtilities.invokeLater { onUpdateDevicesClicked() }
```

Comment:
Shouldn't `SwingUtilities.invokeLater` be moved to the view side? It doesn't look like a part of the logic belonging to the presenter.

Reply:
Removed `invokeLater` entirely and things are working fine. Thanks for catching that